src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs

```diff
@@ -507,6 +507,45 @@ private static bool IsWindowsVersionIncompatibleWithHttp2()
 
             return false;
         }
+
+        internal static SslServerAuthenticationOptions CreateHttp3Options(HttpsConnectionAdapterOptions httpsOptions)
+        {
+            // TODO Set other relevant values on options
+            var sslServerAuthenticationOptions = new SslServerAuthenticationOptions
+            {
+                ServerCertificate = httpsOptions.ServerCertificate,
+                ApplicationProtocols = new List<SslApplicationProtocol>() { new SslApplicationProtocol("h3"), new SslApplicationProtocol("h3-29") },
+                CertificateRevocationCheckMode = httpsOptions.CheckCertificateRevocation ? X509RevocationMode.Online : X509RevocationMode.NoCheck,
+            };
+
+            if (httpsOptions.ServerCertificateSelector != null)
+            {
+                // We can't set both
+                sslServerAuthenticationOptions.ServerCertificate = null;
+                sslServerAuthenticationOptions.ServerCertificateSelectionCallback = (sender, host) =>
+                {
+                    // There is no ConnectionContext available durring the QUIC handshake.
+                    var cert = httpsOptions.ServerCertificateSelector(null, host);
+                    if (cert != null)
+                    {
+                        EnsureCertificateIsAllowedForServerAuth(cert);
+                    }
+                    return cert!;
+                };
+            }
+
+            // DelayCertificate is prohibited by the HTTP/2 and HTTP/3 protocols, ignore it here.
+            if (httpsOptions.ClientCertificateMode == ClientCertificateMode.AllowCertificate
+                    || httpsOptions.ClientCertificateMode == ClientCertificateMode.RequireCertificate)
+            {
+                sslServerAuthenticationOptions.ClientCertificateRequired = true; // This actually means requested, not required.
```

Comment:
Too late to change that name, I take it

Reply:
Yeah, that's years old.

Reply:
Actually, I will update this comment with some recent hard won knowledge. It really means Request and Require, but RemoteCertificateValidationCallback can override that decision, which is what we do for ClientCertificateMode.AllowCertificate.